Resources/public/js/views/services/plugins/ez-locationswapplugin.js

```diff
@@ -0,0 +1,148 @@
+ */
+YUI.add('ez-locationswapplugin', function (Y) {
+    "use strict";
+    /**
+     * Provides the plugin for swap location
+     *
+     * @module ez-locationswapplugin
+     */
+    Y.namespace('eZ.Plugin');
+
+    /**
+     * Swap location plugin. It sets an event handler to swap location
+     *
+     * In order to use it you need to fire `swapLocation` event with parameter
+     * `location` containing the eZ.Location object you want swap with other.
+     *
+     * @namespace eZ.Plugin
+     * @class LocationSwap
+     * @constructor
+     * @extends eZ.Plugin.ViewServiceBase
+     */
+    Y.eZ.Plugin.LocationSwap = Y.Base.create('locationswapplugin', Y.eZ.Plugin.ViewServiceBase, [], {
+
+        initializer: function () {
+            this.onHostEvent('*:swapLocation', this._selectDestinationLocation);
+        },
+
+        /**
+         * swapLocation event handler, launch the universal discovery widget
+         * to choose the location you want to swap with the given one
+         *
+         * @method _selectDestinationLocation
+         * @private
+         * @param {EventFacade} e swapLocation event facade
+         */
+        _selectDestinationLocation: function (e) {
+            var service = this.get('host'),
+                originLocationId = service.get('location').get('id'),
+                originLocationCanHaveChildren = service.get('contentType').get('isContainer'),
+                originLocationHasChildren = originLocationCanHaveChildren &&
+                    (service.get('location').get('childCount') > 0);
+
+            service.fire('contentDiscover', {
+                config: {
+                    title: "Select the location you want to swap with the given one",
+                    contentDiscoveredHandler: Y.bind(this._swapLocations, this),
+                    multiple: false,
+                    isSelectable: function(contentStruct) {
+                        if (originLocationId === contentStruct.location.get('id')) {
+                            return false;
+                        }
+
+                        if (originLocationHasChildren && !contentStruct.contentType.get('isContainer')) {
+                            return false;
+                        }
+
+                        if (!originLocationCanHaveChildren && (contentStruct.location.get('childCount') > 0)) {
+                            return false;
+                        }
+
+                        return true;
+                    },
+                    data: {}
+                },
+            });
+        },
+
+        /**
+         * Swaps the given location with the selected one
+         *
+         * @method _swapLocation
+         * @protected
+         * @param {EventFacade} e
+         */
+        _swapLocations: function (e) {
+            var service = this.get('host'),
+                app = service.get('app'),
+                capi = service.get('capi'),
+                location = service.get('location'),
```

Comment:
since the `swapLocation` event provides a Location model, it would be better to use it instead of retrieving the Location from the service.